### PR TITLE
add receipts for contract-calls/originations

### PR DESCRIPTION
**Pull request by Zett98**
## Problem

We need to know the results of contract origination/calls for indexing.
## Solution

Add receipts for contract origination/ contract calls

### CODE REVIEW

File: src/core_deku/state.ml [hunk 1]
```diff
@@ -9,9 +9,39 @@ type t = {
...
+  initial_tickets : (Ticket_id.t * Amount.t) list;
+}
+
+and receipt =
+  | Receipt_tezos_withdraw       of Ledger.Withdrawal_handle.t
+  | Receipt_contract_origination of {
+      sender : Address.t;
+      outcome : [`Success of contract_origination_changes | `Failure];
+    }
+  | Receipt_contract_call        of {
```
Comment (renatoalencar) [inline src/core_deku/state.ml:34]: This should probably be named `Receipt_contract_invocation` in order to keep naming convention. As well as `contract_invocation_changes`.

File: src/core_deku/state.ml [hunk 2]
```diff
@@ -9,9 +9,39 @@ type t = {
...
+      sender : Address.t;
+      outcome : [`Success of contract_call_changes | `Failure];
+    }
 [@@deriving yojson]
 
+type step = {
```
Comment (renatoalencar) [inline src/core_deku/state.ml:40]: Does step (the name) really makes sense in this context?

File: src/core_deku/state.ml [hunk 3]
```diff
@@ -101,113 +135,176 @@ let rec apply_user_operation ?sender t operation_hash user_operation =
...
-                  acc op)
-              ~init:({ ledger; contract_storage }, None)
-              operations in
-          Ok (receipt, t)) in
-    Ok (t, receipt)
-
-and apply_contract_ops ~source ~sender (t, _) x =
+        let%ok tickets_mapping, tickets, operations =
+          ctx#finalize user_op_list
+          |> Result.map_error ~f:(fun _ ->
```
Comment (renatoalencar) [inline src/core_deku/state.ml:210]: What is the type of `_`? Could you change that to `()` or `_<name>` in order to make it more explicit?

File: src/core_deku/state.ml [hunk 4]
```diff
@@ -101,113 +135,176 @@ let rec apply_user_operation ?sender t operation_hash user_operation =
...
   let report_error error =
     let message =
       match error with
-      | `Origination_error msg -> "Origination error: " ^ msg
-      | `Invocation_error msg -> "Invocation error: " ^ msg
-      | `Insufficient_funds -> "Insufficient funds" in
+      | `Origination_error (msg, _) -> "Origination error: " ^ msg
+      | `Invocation_error (msg, _) -> "Invocation error: " ^ msg
+      | `Insufficient_funds -> "Insufficient funds"
+      | `Invalid_payload _ -> "Invalid contract payload"
```
Comment (renatoalencar) [inline src/core_deku/state.ml:297]: Please, don't ignore error messages and log it back.

**Review by renatoalencar — CHANGES_REQUESTED** [review]

File: src/core_deku/state.ml (hunk 3 above)
Comment (Zett98) [reply src/core_deku/state.ml:210]: This was the receipt, but this code is overwritten now, so I don't get the comment

File: src/core_deku/state.ml (hunk 4 above)
Comment (Zett98) [reply src/core_deku/state.ml:297]: Wdym?

**Review by renatoalencar — APPROVED** [review]
LGTM

**Pull request merged**